### ISSUE TITLE
Can't DROP 'user_id_fk_cache_index' message at the end of update script

### ISSUE THREAD

**Issue by danielh84** [roundcube/roundcubemail#7910]
I updated from 1.4.11 to 1.5-beta

at the end of the bin/install.to script i get:

`Running update script at target...
Executing database schema update.
Updating database schema (2020020100)... [OK]
ERROR: Error in DDL upgrade 2020020101: [1091] Can't DROP 'user_id_fk_cache_index'; check that column/key exists
Updating database schema (2020020101)... [FAILED]
All done.`

On the first look: Everything works fine. No errors. Dark mode looks very nice!

**Comment by alecpl** [comment]
I'm afraid most of that file has not been processed, so you'll have to apply the changes manually.

What MySQL version? Was your 1.4.11 a clean install or an upgrade from an older version (which one)? Could you provide output of `show create table cache_index;` query?

**Comment by danielh84** [comment]
i copied 1.4.11 and the database from the productive environment. The Version was updated from 1.4.10, 1.4.9 and so on... 

I use mySQL-Server 5.7.31-0ubuntu0.18.04.

i get this:

`show create table cache_index

cache_index	CREATE TABLE `cache_index` (
  `user_id` int(10) unsigned NOT NULL,
  `mailbox` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
  `valid` tinyint(1) NOT NULL DEFAULT '0',
  `data` longtext NOT NULL,
  `expires` datetime DEFAULT NULL,
  PRIMARY KEY (`user_id`,`mailbox`),
  KEY `expires_index` (`expires`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8`

Why is "2020020101.sql" executed on update? The file is about 7 month old and should be processed in an earlier update in my opinion or am i wrong?

**Comment by alecpl** [comment]
> i copied 1.4.11 and the database from the productive environment. The Version was updated from 1.4.10, 1.4.9 and so on...

down to what version?

> `show create table cache_index
> 
> cache_index CREATE TABLE `cache_index` (
>  `user_id` int(10) unsigned NOT NULL,
>  `mailbox` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
>  `valid` tinyint(1) NOT NULL DEFAULT '0',
>  `data` longtext NOT NULL,
>  `expires` datetime DEFAULT NULL,
>  PRIMARY KEY (`user_id`,`mailbox`),
>  KEY `expires_index` (`expires`)
> ) ENGINE=InnoDB DEFAULT CHARSET=utf8`

Definitely this table should have a foreign key. https://github.com/roundcube/roundcubemail/blob/34c42f06e1a8b25c022f49ab59f5f26c21c9ec42/SQL/mysql.initial.sql#L68 So, something is wrong with your DB schema.
 
> Why is "2020020101.sql" executed on update? The file is about 7 month old and should be processed in an earlier update in my opinion or am i wrong?

you're wrong.

**Comment by danielh84** [comment]
> > i copied 1.4.11 and the database from the productive environment. The Version was updated from 1.4.10, 1.4.9 and so on...
> 
> down to what version?

I even dont really know... Maybe since 7-8 years. 

> 
> > `show create table cache_index
> > cache_index CREATE TABLE `cache_index` (
> > `user_id` int(10) unsigned NOT NULL,
> > `mailbox` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
> > `valid` tinyint(1) NOT NULL DEFAULT '0',
> > `data` longtext NOT NULL,
> > `expires` datetime DEFAULT NULL,
> > PRIMARY KEY (`user_id`,`mailbox`),
> > KEY `expires_index` (`expires`)
> > ) ENGINE=InnoDB DEFAULT CHARSET=utf8`
> 
> Definitely this table should have a foreign key.
> 
> https://github.com/roundcube/roundcubemail/blob/34c42f06e1a8b25c022f49ab59f5f26c21c9ec42/SQL/mysql.initial.sql#L68
> 
> So, something is wrong with your DB schema.

i checked with 1.4.11. DB schema (productive), there it is also wrong. I cant remember that there was an error in update-process earlier. In the following tables the "CONSTRAINT `user_id_fk_xxxxxxxxxx" line is missing.

- cache_index
- cache_thread
- cache_messages
- contacts
- contactgroups
- contactgroupmembers
- identities
- dictionary
- searches

The "collected_addresses" table is completely missing.

the tables "filestore" and "cache" already includes the "CONSTRAINT...." line. 

i cannot explain what happend here. i looked on older versions of mysql.initial.sql file and there are the contraints since a very long time. I will check my tables and correct them. Dont know why there are no errors in the past.

**Comment by danielh84** [comment]
> the tables "filestore" and "cache" already includes the "CONSTRAINT...." line.

ok, these two tables are right because of

- 2016112200.sql <= created cache table new
- 2018021600.sql / 2018122300.sql <= created filestore

**Comment by danielh84** [comment]
ok, i corrected the structure of all the tables.

Next error was on "2020020101.sql" because of duplicate entries in users table.

Because of (very) old entries in "username" i had users like "Daniel" and "daniel". So i get an error because of duplicate entries in sql-update-query. After deleting the oldest entries (2012-~2015) the update process to 1.5 works fine.

Thanks for your help!

**Comment by alecpl** [comment]
I'm closing this as I don't see an issue to solve. If there was one that caused your db schema "out of sync" it happened old time ago, and we can't track it down.

**Comment by octogonz** [comment]
I encountered this same issue recently. It was while upgrading to RoundCube 1.5 -> 1.6 for an installation whose database dates all the way back to 1.0.0.

For what it's worth, similar to @danielh84 I fixed this by:

1. Export a `mysqldump` of all the old tables.  Make sure to choose the option that includes table names in the SQL insert statements, in case new columns were added.
2. Drop all the old tables 
3. Re-creating the database by sourcing [mysql.initial.sql](https://github.com/roundcube/roundcubemail/blob/a8ce731df933e2f6da28dff0c252b560cbcb89b4/SQL/mysql.initial.sql)
4. Manually insert the 3 important tables: `users`, `identities`, `contacts

Then I re-ran the RoundCube upgrade script, and it succeeded.
